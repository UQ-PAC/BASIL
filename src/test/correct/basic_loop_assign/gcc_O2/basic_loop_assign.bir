000005e1: program
000005af: sub __cxa_finalize(__cxa_finalize_result)
000005e2: __cxa_finalize_result :: out u32 = low:32[R0]

00000353:
0000044b: R16 := 0x20000
00000452: R17 := mem[R16 + 8, el]:u64
00000458: R16 := R16 + 8
0000045d: call R17 with noreturn

000005b0: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005e3: __do_global_dtors_aux_result :: out u32 = low:32[R0]

000002c1:
000002c5: #3 := R31 - 0x20
000002cb: mem := mem with [#3, el]:u64 <- R29
000002d1: mem := mem with [#3 + 8, el]:u64 <- R30
000002d5: R31 := #3
000002db: R29 := R31
000002e3: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002e8: R19 := 0x20000
000002ef: R0 := pad:64[mem[R19 + 0x30]]
000002f5: when 0:0[R0] goto %000002f3
000005d7: goto %0000031c

0000031c:
0000031f: R0 := 0x1F000
00000326: R0 := mem[R0 + 0xFC8, el]:u64
0000032c: when R0 = 0 goto %0000032a
000005d8: goto %00000343

00000343:
00000346: R0 := 0x20000
0000034d: R0 := mem[R0 + 0x28, el]:u64
00000352: R30 := 0x730
00000355: call @__cxa_finalize with return %0000032a

0000032a:
00000332: R30 := 0x734
00000334: call @deregister_tm_clones with return %00000336

00000336:
00000339: R0 := 1
00000341: mem := mem with [R19 + 0x30] <- 7:0[R0]
000005d9: goto %000002f3

000002f3:
000002fd: R19 := mem[R31 + 0x10, el]:u64
00000304: R29 := mem[R31, el]:u64
00000309: R30 := mem[R31 + 8, el]:u64
0000030d: R31 := R31 + 0x20
00000312: call R30 with noreturn

000005b4: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005e4: __libc_start_main_main :: in u64 = R0
000005e5: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005e6: __libc_start_main_arg3 :: in out u64 = R2
000005e7: __libc_start_main_auxv :: in out u64 = R3
000005e8: __libc_start_main_result :: out u32 = low:32[R0]

0000021a:
00000435: R16 := 0x20000
0000043c: R17 := mem[R16, el]:u64
00000442: R16 := R16
00000447: call R17 with noreturn

000005b5: sub _fini(_fini_result)
000005e9: _fini_result :: out u32 = low:32[R0]

00000020:
00000026: #0 := R31 - 0x10
0000002c: mem := mem with [#0, el]:u64 <- R29
00000032: mem := mem with [#0 + 8, el]:u64 <- R30
00000036: R31 := #0
0000003c: R29 := R31
00000043: R29 := mem[R31, el]:u64
00000048: R30 := mem[R31 + 8, el]:u64
0000004c: R31 := R31 + 0x10
00000051: call R30 with noreturn

000005b6: sub _init(_init_result)
000005ea: _init_result :: out u32 = low:32[R0]

00000501:
00000507: #5 := R31 - 0x10
0000050d: mem := mem with [#5, el]:u64 <- R29
00000513: mem := mem with [#5 + 8, el]:u64 <- R30
00000517: R31 := #5
0000051d: R29 := R31
00000522: R30 := 0x590
00000524: call @call_weak_fn with return %00000526

00000526:
0000052b: R29 := mem[R31, el]:u64
00000530: R30 := mem[R31 + 8, el]:u64
00000534: R31 := R31 + 0x10
00000539: call R30 with noreturn

000005b7: sub _start(_start_result)
000005eb: _start_result :: out u32 = low:32[R0]

000001db:
000001e0: R29 := 0
000001e5: R30 := 0
000001eb: R5 := R0
000001f2: R1 := mem[R31, el]:u64
000001f8: R2 := R31 + 8
000001fe: R6 := R31
00000203: R0 := 0x1F000
0000020a: R0 := mem[R0 + 0xFD8, el]:u64
0000020f: R3 := 0
00000214: R4 := 0
00000219: R30 := 0x670
0000021c: call @__libc_start_main with return %0000021e

0000021e:
00000221: R30 := 0x674
00000224: call @abort with return %000005da

000005da:
000005db: call @call_weak_fn with noreturn

000005ba: sub abort()


00000222:
00000477: R16 := 0x20000
0000047e: R17 := mem[R16 + 0x18, el]:u64
00000484: R16 := R16 + 0x18
00000489: call R17 with noreturn

000005bb: sub call_weak_fn(call_weak_fn_result)
000005ec: call_weak_fn_result :: out u32 = low:32[R0]

00000226:
00000229: R0 := 0x1F000
00000230: R0 := mem[R0 + 0xFD0, el]:u64
00000236: when R0 = 0 goto %00000234
000005dc: goto %00000393

00000234:
0000023c: call R30 with noreturn

00000393:
00000396: goto @__gmon_start__

00000394:
00000461: R16 := 0x20000
00000468: R17 := mem[R16 + 0x10, el]:u64
0000046e: R16 := R16 + 0x10
00000473: call R17 with noreturn

000005bd: sub deregister_tm_clones(deregister_tm_clones_result)
000005ed: deregister_tm_clones_result :: out u32 = low:32[R0]

00000242:
00000245: R0 := 0x20000
0000024b: R0 := R0 + 0x30
00000250: R1 := 0x20000
00000256: R1 := R1 + 0x30
0000025c: #1 := ~R0
00000261: #2 := R1 + ~R0
00000267: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000026d: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
00000271: ZF := #2 + 1 = 0
00000275: NF := 63:63[#2 + 1]
0000027b: when ZF goto %00000279
000005dd: goto %00000375

00000375:
00000378: R1 := 0x1F000
0000037f: R1 := mem[R1 + 0xFC0, el]:u64
00000384: when R1 = 0 goto %00000279
000005de: goto %00000388

00000279:
00000281: call R30 with noreturn

00000388:
0000038c: R16 := R1
00000391: call R16 with noreturn

000005c0: sub frame_dummy(frame_dummy_result)
000005ee: frame_dummy_result :: out u32 = low:32[R0]

00000318:
0000031a: call @register_tm_clones with noreturn

000005c1: sub main(main_argc, main_argv, main_result)
000005ef: main_argc :: in u32 = low:32[R0]
000005f0: main_argv :: in out u64 = R1
000005f1: main_result :: out u32 = low:32[R0]

000001a9:
000001ac: R1 := 0x20000
000001b1: R2 := 0x14
000001b6: R0 := 0
000001be: mem := mem with [R1 + 0x34, el]:u32 <- 31:0[R2]
000001c3: call R30 with noreturn

000005c2: sub register_tm_clones(register_tm_clones_result)
000005f2: register_tm_clones_result :: out u32 = low:32[R0]

00000283:
00000286: R0 := 0x20000
0000028c: R0 := R0 + 0x30
00000291: R1 := 0x20000
00000297: R1 := R1 + 0x30
0000029e: R1 := R1 + ~R0 + 1
000002a4: R2 := 0.63:63[R1]
000002ab: R1 := R2 + (R1 ~>> 3)
000002b1: R1 := extend:64[63:1[R1]]
000002b7: when R1 = 0 goto %000002b5
000005df: goto %00000357

00000357:
0000035a: R2 := 0x1F000
00000361: R2 := mem[R2 + 0xFE0, el]:u64
00000366: when R2 = 0 goto %000002b5
000005e0: goto %0000036a

000002b5:
000002bd: call R30 with noreturn

0000036a:
0000036e: R16 := R2
00000373: call R16 with noreturn
