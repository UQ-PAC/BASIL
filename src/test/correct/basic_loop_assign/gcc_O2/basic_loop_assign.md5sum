1a1705720627370be6caf2571048de4c  correct/basic_loop_assign/gcc_O2/a.out
f3496a2984cc911134fc6c41ae4fd723  correct/basic_loop_assign/gcc_O2/basic_loop_assign.adt
898c57621fa9dbef67f7abdcddcf23ea  correct/basic_loop_assign/gcc_O2/basic_loop_assign.bir
829bfa9626b980300c292827c5f31591  correct/basic_loop_assign/gcc_O2/basic_loop_assign.relf
e216c7c532a346c882a342b631284551  correct/basic_loop_assign/gcc_O2/basic_loop_assign.gts
