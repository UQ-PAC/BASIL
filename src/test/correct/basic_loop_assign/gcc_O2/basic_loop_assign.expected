var {:extern} $Gamma_R0: bool;
var {:extern} $Gamma_R1: bool;
var {:extern} $Gamma_R2: bool;
var {:extern} $Gamma_mem: [bv64]bool;
var {:extern} $R0: bv64;
var {:extern} $R1: bv64;
var {:extern} $R2: bv64;
var {:extern} $mem: [bv64]bv8;
const {:extern} $x_addr: bv64;
axiom ($x_addr == 69652bv64);
function {:extern} L(#mem$in: [bv64]bv8, #index: bv64) returns (bool) {
  (if (bvsge64(#index, $x_addr) && bvslt64(#index, bvadd64($x_addr, 4bv64))) then true else false)
}

function {:extern} {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:extern} {:bvbuiltin "bvsge"} bvsge64(bv64, bv64) returns (bool);
function {:extern} {:bvbuiltin "bvsle"} bvsle32(bv32, bv32) returns (bool);
function {:extern} {:bvbuiltin "bvslt"} bvslt32(bv32, bv32) returns (bool);
function {:extern} {:bvbuiltin "bvslt"} bvslt64(bv64, bv64) returns (bool);
function {:extern} gamma_store32(#gammaMap: [bv64]bool, #index: bv64, #value: bool) returns ([bv64]bool) {
  #gammaMap[#index := #value][bvadd64(#index, 1bv64) := #value][bvadd64(#index, 2bv64) := #value][bvadd64(#index, 3bv64) := #value]
}

function {:extern} memory_load32_le(#memory: [bv64]bv8, #index: bv64) returns (bv32) {
  (#memory[bvadd64(#index, 3bv64)] ++ (#memory[bvadd64(#index, 2bv64)] ++ (#memory[bvadd64(#index, 1bv64)] ++ #memory[#index])))
}

function {:extern} memory_load64_le(#memory: [bv64]bv8, #index: bv64) returns (bv64) {
  (#memory[bvadd64(#index, 7bv64)] ++ (#memory[bvadd64(#index, 6bv64)] ++ (#memory[bvadd64(#index, 5bv64)] ++ (#memory[bvadd64(#index, 4bv64)] ++ (#memory[bvadd64(#index, 3bv64)] ++ (#memory[bvadd64(#index, 2bv64)] ++ (#memory[bvadd64(#index, 1bv64)] ++ #memory[#index])))))))
}

function {:extern} memory_store32_le(#memory: [bv64]bv8, #index: bv64, #value: bv32) returns ([bv64]bv8) {
  #memory[#index := #value[8:0]][bvadd64(#index, 1bv64) := #value[16:8]][bvadd64(#index, 2bv64) := #value[24:16]][bvadd64(#index, 3bv64) := #value[32:24]]
}

procedure {:extern} p$rely();
  modifies $Gamma_mem, $mem;
  ensures (forall #i: bv64::  ((($mem[#i] == old($mem[#i])) ==> ($Gamma_mem[#i] == old($Gamma_mem[#i])))));
  ensures (((memory_load32_le($mem, $x_addr) == old(memory_load32_le($mem, $x_addr))) || (bvsle32(memory_load32_le($mem, $x_addr), 10bv32) && bvslt32(old(memory_load32_le($mem, $x_addr)), 10bv32))) || ((memory_load32_le($mem, $x_addr) == 21bv32) && (old(memory_load32_le($mem, $x_addr)) == 20bv32)));
  free ensures (memory_load32_le($mem, 1896bv64) == 131073bv32);
  free ensures (memory_load64_le($mem, 69016bv64) == 1872bv64);
  free ensures (memory_load64_le($mem, 69024bv64) == 1792bv64);
  free ensures (memory_load64_le($mem, 69616bv64) == 1536bv64);
  free ensures (memory_load64_le($mem, 69640bv64) == 69640bv64);

procedure {:extern} p$rely_transitive();
  modifies $Gamma_mem, $mem;
  ensures (((memory_load32_le($mem, $x_addr) == old(memory_load32_le($mem, $x_addr))) || (bvsle32(memory_load32_le($mem, $x_addr), 10bv32) && bvslt32(old(memory_load32_le($mem, $x_addr)), 10bv32))) || ((memory_load32_le($mem, $x_addr) == 21bv32) && (old(memory_load32_le($mem, $x_addr)) == 20bv32)));

implementation {:extern} p$rely_transitive()
{
  call p$rely();
  call p$rely();
}

procedure {:extern} p$rely_reflexive();

implementation {:extern} p$rely_reflexive()
{
  assert (((memory_load32_le($mem, $x_addr) == memory_load32_le($mem, $x_addr)) || (bvsle32(memory_load32_le($mem, $x_addr), 10bv32) && bvslt32(memory_load32_le($mem, $x_addr), 10bv32))) || ((memory_load32_le($mem, $x_addr) == 21bv32) && (memory_load32_le($mem, $x_addr) == 20bv32)));
}

procedure {:extern} p$guarantee_reflexive();
  modifies $Gamma_mem, $mem;

implementation {:extern} p$guarantee_reflexive()
{
  assert (((memory_load32_le($mem, $x_addr) == memory_load32_le($mem, $x_addr)) || ((memory_load32_le($mem, $x_addr) == 20bv32) && (memory_load32_le($mem, $x_addr) == 0bv32))) || ((memory_load32_le($mem, $x_addr) == 20bv32) && bvsle32(memory_load32_le($mem, $x_addr), 10bv32)));
}

procedure p$main_1536();
  modifies $Gamma_R0, $Gamma_R1, $Gamma_R2, $Gamma_mem, $R0, $R1, $R2, $mem;
  requires (memory_load32_le($mem, $x_addr) == 0bv32);
  free requires (memory_load64_le($mem, 69632bv64) == 0bv64);
  free requires (memory_load64_le($mem, 69640bv64) == 69640bv64);
  free requires (memory_load32_le($mem, 1896bv64) == 131073bv32);
  free requires (memory_load64_le($mem, 69016bv64) == 1872bv64);
  free requires (memory_load64_le($mem, 69024bv64) == 1792bv64);
  free requires (memory_load64_le($mem, 69616bv64) == 1536bv64);
  free requires (memory_load64_le($mem, 69640bv64) == 69640bv64);
  ensures ((memory_load32_le($mem, $x_addr) == 20bv32) || (memory_load32_le($mem, $x_addr) == 21bv32));
  free ensures (memory_load32_le($mem, 1896bv64) == 131073bv32);
  free ensures (memory_load64_le($mem, 69016bv64) == 1872bv64);
  free ensures (memory_load64_le($mem, 69024bv64) == 1792bv64);
  free ensures (memory_load64_le($mem, 69616bv64) == 1536bv64);
  free ensures (memory_load64_le($mem, 69640bv64) == 69640bv64);

implementation p$main_1536()
{
  var #x_old: bv32;
  b#main_entry:
    assume {:captureState "main_entry"} true;
    $R1, $Gamma_R1 := 69632bv64, true;
    $R2, $Gamma_R2 := 20bv64, true;
    $R0, $Gamma_R0 := 0bv64, true;
    call p$rely();
    #x_old := memory_load32_le($mem, $x_addr);
    assert (L($mem, bvadd64($R1, 20bv64)) ==> $Gamma_R2);
    $mem, $Gamma_mem := memory_store32_le($mem, bvadd64($R1, 20bv64), $R2[32:0]), gamma_store32($Gamma_mem, bvadd64($R1, 20bv64), $Gamma_R2);
    assert (((memory_load32_le($mem, $x_addr) == #x_old) || ((memory_load32_le($mem, $x_addr) == 20bv32) && (#x_old == 0bv32))) || ((memory_load32_le($mem, $x_addr) == 20bv32) && bvsle32(#x_old, 10bv32)));
    assume {:captureState "%000001bd"} true;
    goto b#main_basil_return_1;
  b#main_basil_return_1:
    assume {:captureState "main_basil_return_1"} true;
    return;
}

